server/service/apple_mdm.go

```diff
@@ -215,16 +213,80 @@ func getMDMAppleCommandResultsEndpoint(ctx context.Context, request interface{},
 	}, nil
 }
 
-func (svc *Service) GetMDMAppleCommandResults(ctx context.Context, commandUUID string) (map[string]*fleet.MDMAppleCommandResult, error) {
-	if err := svc.authz.Authorize(ctx, &fleet.MDMAppleCommandResult{}, fleet.ActionRead); err != nil {
+func (svc *Service) GetMDMAppleCommandResults(ctx context.Context, commandUUID string) ([]*fleet.MDMAppleCommandResult, error) {
+	// first, authorize that the user has the right to list hosts
+	if err := svc.authz.Authorize(ctx, &fleet.Host{}, fleet.ActionList); err != nil {
 		return nil, ctxerr.Wrap(ctx, err)
 	}
 
+	vc, ok := viewer.FromContext(ctx)
+	if !ok {
+		return nil, fleet.ErrNoContext
+	}
+
+	// check that command exists first, to return 404 on invalid commands
+	// (the command may exist but have no results yet).
+	if _, err := svc.ds.GetMDMAppleCommandRequestType(ctx, commandUUID); err != nil {
+		return nil, err
+	}
+
+	// next, we need to read the command results before we know what hosts (and
+	// therefore what teams) we're dealing with.
 	results, err := svc.ds.GetMDMAppleCommandResults(ctx, commandUUID)
 	if err != nil {
 		return nil, err
 	}
 
+	// now we can load the hosts (lite) corresponding to those command results,
+	// and do the final authorization check with the proper team(s). Include observers,
+	// as they are able to view command results for their teams' hosts.
+	filter := fleet.TeamFilter{User: vc.User, IncludeObserver: true}
+	hostUUIDs := make([]string, len(results))
+	for i, res := range results {
+		hostUUIDs[i] = res.DeviceID
+	}
+	hosts, err := svc.ds.ListHostsLiteByUUIDs(ctx, filter, hostUUIDs)
+	if err != nil {
+		return nil, err
+	}
+	if len(hosts) == 0 {
+		// do not return 404 here, as it's possible for a command to not have
+		// results yet
+		return nil, nil
+	}
+
+	// collect the team IDs and verify that the user has access to run commands
+	// on all affected teams. Index the hosts by uuid for easly lookup as
+	// afterwards we'll want to store the hostname on the returned results.
+	hostsByUUID := make(map[string]*fleet.Host, len(hosts))
+	teamIDs := make(map[uint]bool)
+	for _, h := range hosts {
+		var id uint
+		if h.TeamID != nil {
+			id = *h.TeamID
+		}
+		teamIDs[id] = true
+		hostsByUUID[h.UUID] = h
+	}
+
+	var command fleet.MDMAppleCommand
+	for tmID := range teamIDs {
+		command.TeamID = &tmID
+		if tmID == 0 {
+			command.TeamID = nil
+		}
+
+		if err := svc.authz.Authorize(ctx, command, fleet.ActionRead); err != nil {
```

Comment:
is there an edge case lurking here? I imagine a host switching teams and this endpoint returning an unauthorized response for all the results.

I don't know if there's a better solution, but raising just in case.

Reply:
Yeah, what I 'm thinking is that we could return partial results, but there's currently no way to indicate that this was the case so this could be surprising/confusing. On the other hand, as you mention, there's the risk that one targeted host gets moved and it blocks getting those results for non-global users.

@noahtalerman any thoughts about that? What happens here is that this is where the endpoint checks if the user is authorized to view results for the team of the host affected by the MDM command. The edge case here is that if a command targeted 2 hosts and one of them changes team, then the user that can only see one of those teams would receive an authorization error instead of seeing the results (as they don't have access to one of the teams involved).

An option would be to return partial results, but we'd probably need a way to indicate that this was the case in the `fleetctl get mdm-command-results` output (so that the user is aware that there were more results for that command, but they're not authorized to see them). 

I don't think this is critical at the moment, as the `fleetctl mdm run-command` command can only target one host at a time (so as not to delay the release), but definitely something to think about and possibly modify in the future?

Reply:
2cents: I agree on tackling this as a future thing, not worth blocking this PR. It just requires some thought and as Martin mentions is not critical at the moment.

Reply:
@noahtalerman  The `fleetctl` command does target a single host, but the API and Apple's MDM support targeting multiple hosts with a single command - and we do this internally too, so once we have the "list all commands executed" feature, it will be possible to view commands that were run on multiple hosts (which I assumed was also why we print a table of results in `get mdm-command-results`, as there may be multiple results for a single command).

Reply:
> we do this internally too

> once we have the "list all commands executed" feature, it will be possible to view commands that were run on multiple hosts

Oh, ok.

@mna if I'm understanding correctly, it will be possible only if a user used the Fleet API to run the command? It won't be possible if the user used the `fleetctl` command, right?

Reply:
@noahtalerman 

> It won't be possible if the user used the fleetctl command, right?

Correct, `fleetctl mdm run-command` can only target a single host. 

> this will be possible only if a user used the Fleet API to run the command?

Not _only_ the API, as Fleet internally runs commands to install/remove profiles and other administrative tasks (I guess "install enterprise application" would also be recorded as a command) and unless we do something special to filter them out, those commands would be printed in the output of `fleetctl get mdm-commands` (the list of all commands executed).

Reply:
@noahtalerman 

> those commands would be printed in the output of fleetctl get mdm-commands

to complete my thought on this, once it's printed in that list, nothing prevents the user from copying a command id and running `fleetctl get mdm-command-results --id <id>` to view the result of that command, which may have targeted multiple hosts.

Reply:
> Not only the API, as Fleet internally runs commands to install/remove profiles and other administrative tasks

Ah, that's right. Yeah we decided to not do any filtering out of the internal commands for now.

Ok, thanks for walking me through that, I think I have the full picture.

> the user that can only see one of those teams would receive an authorization error instead of seeing the results (as they don't have access to one of the teams involved).

@mna is this the current behavior? I think this is ok. 

Maybe we should update the authorization error message so the user knows what's happening: they don't have access to one or more hosts.

What do you think?

Reply:
@noahtalerman 

> Ok, thanks for walking me through that, I think I have the full picture.

No problem! Those are pretty specific details.

> is this the current behavior? I think this is ok.

Yes, it is.

> Maybe we should update the authorization error message so the user knows what's happening: they don't have access to one or more hosts.

That's a great idea, I'll push a new PR with that improvement, thanks!

Reply:
https://github.com/fleetdm/fleet/pull/10997

Reply:
Sweet!